File: src/lib/github/actions.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { invoke } from "@/lib/tauri/invoke";

// ── Types (mirror the Rust structs in github/actions.rs) ─────────────────────

export interface WorkflowRun {
  id: number;
  number: number;
  displayTitle: string;
  /** queued | in_progress | completed | waiting | requested | pending */
  status: string;
  /** success | failure | cancelled | skipped | … ; "" while still running */
  conclusion: string;
  workflowName: string;
  headBranch: string;
  event: string;
  createdAt: string;
  /** When the run started executing (after queue); "" if never started. */
  startedAt: string;
  updatedAt: string;
  url: string;
  headSha: string;
}

export interface RunStep {
  name: string;
  status: string;
  conclusion: string;
  number: number;
  startedAt: string;
  completedAt: string;
}

export interface RunJob {
  id: number;
  name: string;
  status: string;
  conclusion: string;
  startedAt: string;
  completedAt: string;
  url: string;
  steps: RunStep[];
}

export interface RunDetail {
  id: number;
  number: number;
  displayTitle: string;
  status: string;
  conclusion: string;
  workflowName: string;
  headBranch: string;
  event: string;
  createdAt: string;
  url: string;
  headSha: string;
  jobs: RunJob[];
}

export interface Workflow {
  id: number;
  name: string;
  path: string;
  /** active | disabled_manually | disabled_inactivity */
  state: string;
}

// ── Status helpers ───────────────────────────────────────────────────────────

const ACTIVE_STATUSES = new Set([
  "queued",
  "in_progress",
  "waiting",
  "requested",
  "pending",
]);

/** A run/job still executing (so the UI keeps polling and offers Cancel). */
export const isRunActive = (status: string) => ACTIVE_STATUSES.has(status);

// ── API wrappers ─────────────────────────────────────────────────────────────
//
// Reads AND writes go through the provider-neutral `forge_ci_*` commands (GitHub
// via `gh run …`, GitLab via `glab` pipelines → the same `WorkflowRun`/`RunDetail`
// shapes; re-run / cancel / dispatch dispatch per provider too). Only the
// workflow list stays `gh_*` — GitLab has no workflow analogue (one `.gitlab-ci.yml`
// per project), so its dispatch is ref+variables with no workflow picker.

export const forgeCiRunList = (
  repoPath: string,
  limit: number,
  branch?: string,
) =>
  invoke<WorkflowRun[]>("forge_ci_run_list", {
    repoPath,
    limit,
    branch: branch?.trim() || null,
  });

export const forgeCiRunView = (repoPath: string, runId: number) =>
  invoke<RunDetail>("forge_ci_run_view", { repoPath, runId });

export const forgeCiRunRerun = (
  repoPath: string,
  runId: number,
  failed: boolean,
) => invoke<void>("forge_ci_run_rerun", { repoPath, runId, failed });

export const forgeCiRunCancel = (repoPath: string, runId: number) =>
  invoke<void>("forge_ci_run_cancel", { repoPath, runId });

export const forgeCiRunFailedLogs = (repoPath: string, runId: number) =>
  invoke<string>("forge_ci_run_failed_logs", { repoPath, runId });

/** One job's failed-step logs (fallback: full job log), for AI debugging. */
export const forgeCiJobLogs = (repoPath: string, jobId: number) =>
  invoke<string>("forge_ci_job_logs", { repoPath, jobId });

export const ghWorkflowList = (repoPath: string) =>
  invoke<Workflow[]>("gh_workflow_list", { repoPath });

/** Start a run: GitHub dispatches `workflow` on the ref with `inputs`; GitLab runs
 *  a new pipeline on the ref with `inputs` as variables (send `workflow` empty). */
export const forgeCiDispatch = (
  repoPath: string,
  workflow: string,
  gitRef: string,
  inputs: Record<string, string>,
) => invoke<void>("forge_ci_dispatch", { repoPath, workflow, gitRef, inputs });

// ── Queries ──────────────────────────────────────────────────────────────────

/** Polls every 5s while any listed run is active, otherwise stays idle. */
export function useWorkflowRuns(
  repo: string,
  enabled: boolean,
  branch?: string,
) {
  return useQuery({
    queryKey: ["repo", repo, "actions", "runs", branch ?? ""] as const,
    queryFn: () => forgeCiRunList(repo, 40, branch),
    enabled,
    staleTime: 10_000,
    refetchInterval: (query) =>
      (query.state.data ?? []).some((r) => isRunActive(r.status))
        ? 5000
        : false,
  });
}

export function useRunDetail(repo: string, runId: number | null) {
  return useQuery({
    queryKey: ["repo", repo, "actions", "run", runId ?? 0] as const,
    queryFn: () => forgeCiRunView(repo, runId ?? 0),
    enabled: runId !== null,
    refetchInterval: (query) =>
      query.state.data && isRunActive(query.state.data.status) ? 5000 : false,
  });
}

/**
 * The single most recent run on a branch, for the header CI badge. Polls fast
 * while it's active, slowly otherwise so a freshly-pushed run still shows up.
 */
export function useLatestRun(
  repo: string,
  enabled: boolean,
  branch: string | null,
) {
  return useQuery({
    queryKey: ["repo", repo, "actions", "latest", branch ?? ""] as const,
    queryFn: async () =>
      (await forgeCiRunList(repo, 1, branch ?? undefined))[0] ?? null,
    enabled: enabled && Boolean(branch),
    staleTime: 15_000,
    refetchInterval: (query) =>
      query.state.data && isRunActive(query.state.data.status) ? 8000 : 30_000,
  });
}

export function useWorkflows(repo: string, enabled: boolean) {
  return useQuery({
    queryKey: ["repo", repo, "actions", "workflows"] as const,
    queryFn: () => ghWorkflowList(repo),
    enabled,
    staleTime: 5 * 60_000,
  });
}

/** Failed-step logs, fetched only when the user expands them. */
export function useRunFailedLogs(
  repo: string,
  runId: number | null,
  enabled: boolean,
) {
  return useQuery({
    queryKey: ["repo", repo, "actions", "run", runId ?? 0, "logs"] as const,
    queryFn: () => forgeCiRunFailedLogs(repo, runId ?? 0),
    enabled: enabled && runId !== null,
    staleTime: 30_000,
  });
}

/** One job's logs (failed steps, or the full log), fetched only when expanded. */
export function useJobLogs(
  repo: string,
  jobId: number | null,
  enabled: boolean,
) {
  return useQuery({
    queryKey: ["repo", repo, "actions", "job", jobId ?? 0, "logs"] as const,
    queryFn: () => forgeCiJobLogs(repo, jobId ?? 0),
    enabled: enabled && jobId !== null,
    staleTime: 30_000,
  });
}

// ── Mutations ────────────────────────────────────────────────────────────────

function useActionsMutation<TArgs>(
  repo: string,
  mutationFn: (args: TArgs) => Promise<void>,
) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn,
    // Scope the refresh to Actions — re-run/cancel/dispatch don't touch git state.
    onSettled: () =>
      queryClient.invalidateQueries({
        queryKey: ["repo", repo, "actions"],
      }),
  });
}

export function useRerunRun(repo: string) {
  return useActionsMutation(repo, (args: { runId: number; failed: boolean }) =>
    forgeCiRunRerun(repo, args.runId, args.failed),
  );
}

export function useCancelRun(repo: string) {
  return useActionsMutation(repo, (runId: number) =>
    forgeCiRunCancel(repo, runId),
  );
}

export function useRunWorkflow(repo: string) {
  return useActionsMutation(
    repo,
    (args: {
      workflow: string;
      gitRef: string;
      inputs: Record<string, string>;
    }) => forgeCiDispatch(repo, args.workflow, args.gitRef, args.inputs),
  );
}

File: src-tauri/src/github/actions.rs
//! GitHub Actions support, driven through the `gh` CLI (`gh run …` /
//! `gh workflow …`). Read views list and inspect workflow runs; mutations
//! re-run, cancel, and manually dispatch workflows. All repo-scoped via the
//! working directory, like the PR commands.

use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use crate::error::{AppError, AppResult};
use crate::github::runner::{run_gh, run_gh_raw, GH_NETWORK_TIMEOUT};

/// gh emits `null` for a not-yet-decided conclusion (and timestamps that
/// haven't happened); fold those into "" so the frontend sees a plain string.
fn de_null_string<'de, D>(d: D) -> Result<String, D::Error>
where
    D: serde::Deserializer<'de>,
{
    Ok(Option::<String>::deserialize(d)?.unwrap_or_default())
}

fn validate_ref(name: &str) -> AppResult<()> {
    if name.is_empty() || name.starts_with('-') {
        return Err(AppError::InvalidArgument(format!("invalid ref: {name}")));
    }
    Ok(())
}

/// One workflow run in the list view.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WorkflowRun {
    #[serde(rename(serialize = "id", deserialize = "databaseId"))]
    pub id: u64,
    #[serde(default)]
    pub number: u64,
    #[serde(default)]
    pub display_title: String,
    /// "queued" | "in_progress" | "completed" | "waiting" | "requested" | "pending".
    #[serde(default)]
    pub status: String,
    /// "success" | "failure" | "cancelled" | … ; "" while still running.
    #[serde(default, deserialize_with = "de_null_string")]
    pub conclusion: String,
    #[serde(default)]
    pub workflow_name: String,
    #[serde(default)]
    pub head_branch: String,
    #[serde(default)]
    pub event: String,
    #[serde(default)]
    pub created_at: String,
    #[serde(default)]
    pub updated_at: String,
    #[serde(default)]
    pub url: String,
    #[serde(default)]
    pub head_sha: String,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RunStep {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub status: String,
    #[serde(default, deserialize_with = "de_null_string")]
    pub conclusion: String,
    #[serde(default)]
    pub number: i64,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RunJob {
    #[serde(rename(serialize = "id", deserialize = "databaseId"))]
    pub id: u64,
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub status: String,
    #[serde(default, deserialize_with = "de_null_string")]
    pub conclusion: String,
    #[serde(default, deserialize_with = "de_null_string")]
    pub started_at: String,
    #[serde(default, deserialize_with = "de_null_string")]
    pub completed_at: String,
    #[serde(default)]
    pub url: String,
    #[serde(default)]
    pub steps: Vec<RunStep>,
}

/// A run plus its jobs/steps, for the detail view.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RunDetail {
    #[serde(rename(serialize = "id", deserialize = "databaseId"))]
    pub id: u64,
    #[serde(default)]
    pub number: u64,
    #[serde(default)]
    pub display_title: String,
    #[serde(default)]
    pub status: String,
    #[serde(default, deserialize_with = "de_null_string")]
    pub conclusion: String,
    #[serde(default)]
    pub workflow_name: String,
    #[serde(default)]
    pub head_branch: String,
    #[serde(default)]
    pub event: String,
    #[serde(default)]
    pub created_at: String,
    #[serde(default)]
    pub url: String,
    #[serde(default)]
    pub head_sha: String,
    #[serde(default)]
    pub jobs: Vec<RunJob>,
}

/// A repo workflow, for the "Run workflow" picker.
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Workflow {
    pub id: u64,
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub path: String,
    /// "active" | "disabled_manually" | "disabled_inactivity".
    #[serde(default)]
    pub state: String,
}

const RUN_LIST_FIELDS: &str = "databaseId,number,displayTitle,status,conclusion,workflowName,headBranch,event,createdAt,updatedAt,url,headSha";
const RUN_VIEW_FIELDS: &str = "databaseId,number,displayTitle,status,conclusion,workflowName,headBranch,event,createdAt,url,headSha,jobs";
/// Failed-step logs can run to many MB; keep the tail (failures land at the end).
const RUN_LOG_CAP: usize = 200_000;

/// Recent workflow runs, newest first; optionally scoped to one branch.
#[tauri::command]
pub async fn gh_run_list(
    repo_path: String,
    limit: u32,
    branch: Option<String>,
) -> AppResult<Vec<WorkflowRun>> {
    let limit = limit.clamp(1, 100).to_string();
    let mut args: Vec<&str> = vec![
        "run",
        "list",
        "-L",
        limit.as_str(),
        "--json",
        RUN_LIST_FIELDS,
    ];
    if let Some(b) = branch.as_deref().filter(|s| !s.is_empty()) {
        validate_ref(b)?;
        args.push("--branch");
        args.push(b);
    }
    let out = run_gh(Some(&repo_path), &args, GH_NETWORK_TIMEOUT).await?;
    serde_json::from_str(&out.stdout_lossy())
        .map_err(|e| AppError::Gh(format!("could not parse gh run list: {e}")))
}

/// One run with its jobs and steps.
#[tauri::command]
pub async fn gh_run_view(repo_path: String, run_id: u64) -> AppResult<RunDetail> {
    let id = run_id.to_string();
    let out = run_gh(
        Some(&repo_path),
        &["run", "view", &id, "--json", RUN_VIEW_FIELDS],
        GH_NETWORK_TIMEOUT,
    )
    .await?;
    serde_json::from_str(&out.stdout_lossy())
        .map_err(|e| AppError::Gh(format!("could not parse gh run view: {e}")))
}

/// Re-runs a completed run — all jobs, or only the failed ones.
#[tauri::command]
pub async fn gh_run_rerun(repo_path: String, run_id: u64, failed: bool) -> AppResult<()> {
    let id = run_id.to_string();
    let mut args = vec!["run", "rerun", id.as_str()];
    if failed {
        args.push("--failed");
    }
    run_gh(Some(&repo_path), &args, GH_NETWORK_TIMEOUT).await?;
    Ok(())
}

/// Cancels an in-progress run.
#[tauri::command]
pub async fn gh_run_cancel(repo_path: String, run_id: u64) -> AppResult<()> {
    let id = run_id.to_string();
    run_gh(
        Some(&repo_path),
        &["run", "cancel", &id],
        GH_NETWORK_TIMEOUT,
    )
    .await?;
    Ok(())
}

/// Logs of only the failed steps (`gh run view --log-failed`), tail-capped.
/// Read raw because gh exits non-zero on a failed run.
#[tauri::command]
pub async fn gh_run_failed_logs(repo_path: String, run_id: u64) -> AppResult<String> {
    let id = run_id.to_string();
    let out = run_gh_raw(
        Some(&repo_path),
        &["run", "view", &id, "--log-failed"],
        GH_NETWORK_TIMEOUT,
    )
    .await?;
    let mut text = out.stdout_lossy();
    if text.trim().is_empty() {
        // No failed-step logs (e.g. cancelled, or still running) — surface gh's
        // own message instead of an empty pane.
        text = out.stderr.trim().to_string();
    }
    if text.len() > RUN_LOG_CAP {
        let mut start = text.len() - RUN_LOG_CAP;
        while !text.is_char_boundary(start) {
            start += 1;
        }
        text = format!("…(earlier output truncated)\n{}", &text[start..]);
    }
    Ok(text)
}

/// Logs for one job, for AI debugging. Prefers the failed-step logs (highest
/// signal); falls back to the full job log when gh returns nothing for
/// `--log-failed`. Tighter cap than the run-level logs since this is fed to a
/// model. Read raw because gh exits non-zero on a failed run.
const JOB_LOG_CAP: usize = 60_000;

#[tauri::command]
pub async fn gh_job_logs(repo_path: String, job_id: u64) -> AppResult<String> {
    let id = job_id.to_string();
    let mut out = run_gh_raw(
        Some(&repo_path),
        &["run", "view", "--job", &id, "--log-failed"],
        GH_NETWORK_TIMEOUT,
    )
    .await?;
    let mut text = out.stdout_lossy();
    if text.trim().is_empty() {
        out = run_gh_raw(
            Some(&repo_path),
            &["run", "view", "--job", &id, "--log"],
            GH_NETWORK_TIMEOUT,
        )
        .await?;
        text = out.stdout_lossy();
    }
    if text.trim().is_empty() {
        text = out.stderr.trim().to_string();
    }
    if text.len() > JOB_LOG_CAP {
        let mut start = text.len() - JOB_LOG_CAP;
        while !text.is_char_boundary(start) {
            start += 1;
        }
        text = format!("…(earlier output truncated)\n{}", &text[start..]);
    }
    Ok(text)
}

/// The repo's workflows, for the manual-dispatch picker.
#[tauri::command]
pub async fn gh_workflow_list(repo_path: String) -> AppResult<Vec<Workflow>> {
    let out = run_gh(
        Some(&repo_path),
        &["workflow", "list", "--all", "--json", "id,name,path,state"],
        GH_NETWORK_TIMEOUT,
    )
    .await?;
    serde_json::from_str(&out.stdout_lossy())
        .map_err(|e| AppError::Gh(format!("could not parse gh workflow list: {e}")))
}

/// Manually dispatches a workflow (`workflow_dispatch`) on a ref, with inputs.
/// `workflow` is the numeric id or the file name (e.g. "ci.yml").
#[tauri::command]
pub async fn gh_workflow_run(
    repo_path: String,
    workflow: String,
    git_ref: String,
    inputs: HashMap<String, String>,
) -> AppResult<()> {
    if workflow.is_empty() || workflow.starts_with('-') {
        return Err(AppError::InvalidArgument(format!(
            "invalid workflow: {workflow}"
        )));
    }
    validate_ref(&git_ref)?;
    let mut args: Vec<String> = vec![
        "workflow".into(),
        "run".into(),
        workflow,
        "--ref".into(),
        git_ref,
    ];
    for (k, v) in &inputs {
        if k.is_empty() || k.starts_with('-') {
            return Err(AppError::InvalidArgument(format!("invalid input key: {k}")));
        }
        args.push("-f".into());
        args.push(format!("{k}={v}"));
    }
    let arg_refs: Vec<&str> = args.iter().map(String::as_str).collect();
    run_gh(Some(&repo_path), &arg_refs, GH_NETWORK_TIMEOUT).await?;
    Ok(())
}
